### PR TITLE
[Android] Fix audio engine initialisation on startup

### PR DESCRIPTION
**Pull request by hugbug**
## Description

If screen resolution or refresh rate is adjusted during Kodi startup (as defined in Kodi settings), the audio engine will be reinitialised after the HDMI handshake is completed.

## Motivation and context

Kodi has a setting for screen resolution and refresh rate (Settings -> System -> Display). On startup Kodi may need to change the resolution or refresh rate to match the settings. This causes HDMI handshake, which may take several seconds. However the startup routine continues. It may happen that during initialisation of audio engine the handshake has not been completed yet. In that case the audio capabilities and audio device list in audio engine may be incorrect.

In particular this happens on my Fire TV Stick 4K Max (both 1st gen and 2nd gen) connected to AVR. If I change refresh rate in Kodi settings to 23.97 fps, then on next Kodi start the Kodi IEC packer disappears, causing all kind of issues.

The problem was also reported on forum - [Fire TV 4K max V2 Issues](https://forum.kodi.tv/showthread.php?tid=375529&pid=3176958#pid3176958) (s. posts No. 4 and 24).

### Implementation details
When display mode is changed the event ACTION_HDMI_AUDIO_PLUG is sent to the application. However all events were ignored until Kodi was fully initialised. The PR makes a small change to how events are processed during startup process:
- event ACTION_HDMI_AUDIO_PLUG can now be processed even if the app isn't fully initialised yet.
- one event ACTION_HDMI_AUDIO_PLUG is always sent, even if the display mode wasn't changed. To react only on event occurred after display mode change we use variable (class member) `m_wakeUp`, which we set in `onDisplayChanged`. I think it's appropriate to use this variable instead of introducing another one, because the purpose is very similar to the first use case of the variable.
- a small change in logging was made. Instead of `CLog::Log` we now use `android_printf` - a function used throughout `CXBMCApp` during startup, when `CLog` may not be initialised yet.

## How has this been tested?

Tested on Fire TV Stick 4K Max 2nd gen, connected to AVR Denon AVC-X4800H, connected to TV LG C1. Test procedure:

1. In Kodi Settings under Settings -> System -> Display choose a screen resolution or refresh rate, which is different from the Fire TV System settings. For example set refresh rate to 23.97.
2. Restart Kodi.
3. Go to Kodi Settings -> System -> Audio and check available audio device. Is Kodi IEC packer listed there?

### Test results
**Before fix:** On my device on step 3 the audio device list had only Android IEC packer. Audio passthrough, which were configured via Kodi IEC packer, didn't work anymore.

**After fix:** Kodi IEC packer is available and selected in audio device list. Audio passthrough works as expected.

## What is the effect on users?

Users who were having problem with audio passthrough when using custom resolution or refresh rate in Kodi should have the issue fixed. In particular Fire TV devices connected to AVR are known to be affected.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by fritsch** [comment]
The printfs you use, cause logger might not be available yet, right?

In general I agree what you are doing there and it is sane. As I have limited time for kodi right now in general I kindly want to ask @thexai and @joseluismarti for their thoughts. Thanks in advance.

**Comment by malard** [comment]
you've just saved me a bunch of time as I was about to start debugging this exact issue :)

**Comment by hugbug** [comment]
@malard, do you also have Fire TV or some other device and is it also connected to an AVR (which model) or directly to a TV?

**Comment by hugbug** [comment]
Reinitialisation of audio engine after each refresh rate change would be an overkill. It easily can be done in ACTION_HDMI_AUDIO_PLUG event - just do audio engine reset every time the event with parameter "cable plugged-in" is received. For this it is enough to remove check of `m_wakeUp` from https://github.com/xbmc/xbmc/blob/d5e1c8787efae505ee02bf23c6ecbdbfc648b952/xbmc/platform/android/activity/XBMCApp.cpp#L1336

However, at least on my system, that adds a delay of up to 500ms when starting video playback (I tried this). That's not a good thing too.

I hope current PR in combination with #24369 eliminate most issues without adding unnecessary delays. If it doesn't - we still can try the general approach with reset after each ACTION_HDMI_AUDIO_PLUG event.

**Comment by joseluismarti** [comment]
> In general I agree what you are doing there and it is sane. As I have limited time for kodi right now in general I kindly want to ask @thexai and @joseluismarti for their thoughts. Thanks in advance.

Hi, I don't have an AVR device and without testing and seeing if any problems arise I can't help much, sorry.

**Comment by CastagnaIT** [comment]
> in particular this happens on my Fire TV Stick 4K Max (both 1st gen and 2nd gen) connected to AVR. If I change refresh rate in Kodi settings to 23.97 fps, then on next Kodi start the Kodi IEC packer disappears, causing all kind of issues.

i can confirm that this problem can happens also on nvidia shield where Audio passthrough is full lost until you force close app, i thought it was a problem with my old amplifier but good to see that isnt it,
i can try to see if im able reproduce it again by following mentioned steps

**Comment by the-black-eagle** [comment]
I can confirm that this solves the issue of the disappearing Kodi IEC packer on Firestick 4K Max 2 when Kodi starts in a different resolution/refresh rate to the firestick.

Build failure is unrelated.

**Comment by thexai** [comment]
I tried to reproduce this in my setup without success: changed GUI refresh rate to 23.967 Hz and restarted. Kodi IEC packer is available same as before and audio passthrough works same as before...

Here Shield Pro 2019 (FW 9.2) + Denon AVR-X1600H + Sony AG9 TV + CEC enabled on all devices and zero issues related to audio or video (passthrough + sync display refresh rate enabled).

Anyway seems now everybody has same issue then probably changes makes sense....

There is any bug issue opened for this?

File: xbmc/platform/android/activity/XBMCApp.cpp [hunk 1]
```diff
@@ -1271,11 +1271,18 @@ void CXBMCApp::SetSystemVolume(float percent)
 
 void CXBMCApp::onReceive(CJNIIntent intent)
 {
-  if (!g_application.IsInitialized())
+  std::string action = intent.getAction();
+  android_printf("CXBMCApp::onReceive - Got intent. Action: %s", action.c_str());
+
+  // Most actions can be processed only after the app is fully initialized,
+  // but some actions should be processed even during initilization phase.
+  if (!g_application.IsInitialized() && !(action == CJNIAudioManager::ACTION_HDMI_AUDIO_PLUG))
```
Comment (thexai) [inline xbmc/platform/android/activity/XBMCApp.cpp:1279]:
```suggestion
  if (!g_application.IsInitialized() && action != CJNIAudioManager::ACTION_HDMI_AUDIO_PLUG)
```

It seems to be equivalent and a more common code

**Comment by frodo19** [comment]
Yes I confirm too, this fixed the issue in Firestick 4K max 2nd.
Now I can set my fsmax2 to 1920*1080 59,94, and Kodi IEC packer return after reboot.

**Comment by hugbug** [comment]
> There is any bug issue opened for this?

I don't think so. I've seen problem reports on forum only (topic linked in the first post).

File: xbmc/platform/android/activity/XBMCApp.cpp (hunk 1 above)
Comment (hugbug) [reply xbmc/platform/android/activity/XBMCApp.cpp]: Thank you.

**Review by thexai — APPROVED** [review]
Let's see what happens...

**Comment by CrystalP** [comment]
Speaking of common code, isn't there a lost/reset display mechanism to reinitialize the sinks? Would it work here?

**Comment by fritsch** [comment]
> Speaking of common code, isn't there a lost/reset display mechanism to reinitialize the sinks? Would it work here?

Normally yes. But we have a different situation here. We have a "mainApp" that is doing the Java / intent / whatever things. Here we receive the intents and try to forward them to the subsystems inside kodi.

Kodi itself has no mechanism to queue such events durch startup, they are lost or reach someone that cannot do anything cause his IsServiceAvailable() or ptr check equals to false / off.

The LOST_DEVICE infrastructure only works after full initialization only as well. 

In general it might be benefitial to enqueue those events and just pop them out of a queue at the end of our Initialization. While - sadly - we also rely in Android on some events synchroneously as we trigger them ourselves.

Which is why I think this approach here is currently the least intrusive one, while it only fixes Audio initialization, but nothingt (potential) else.

**Review by fritsch — APPROVED** [review]

**Comment by fritsch** [comment]
I would push it if you agree @CrystalP

**Comment by thexai** [comment]
Yes but happens that all events are ignored until app is full initialized:

https://github.com/xbmc/xbmc/blob/c0044d4fb19c59f747f0bdcd597ad3b7b8a6c34d/xbmc/platform/android/activity/XBMCApp.cpp#L1272-L1275

And this PR changes this...

**Comment by CrystalP** [comment]
> I would push it if you agree @CrystalP

sure go ahead.

**Comment by fritsch** [comment]
Thanks again for your contribution @hugbug - keep them coming :-)

**Comment by CastagnaIT** [comment]
i didn't make it in time test yesterday
anyway i tested now and i was not able to reproduce the problem manually, take into consideration that with my shield this problem happens very rarely 1 times in one or two months and coincidence that it happened last week, tested this PR and not found problems for now

**Comment by fritsch** [comment]
Thanks much @CastagnaIT - awesome.

**Pull request merged**